kernel/src/scan/file_stream.rs

```diff
@@ -142,9 +162,106 @@ impl LogReplayScanner {
+            Some(ref filter_vector) => filter_vector.clone(),
+            None => vec![false; actions.length()],
+        };
+
+        let visitor = self.setup_batch_process(filter_vector, actions, is_log_batch)?;
+
+        for (add, index) in visitor.adds.into_iter() {
+            // Note: each (add.path + add.dv_unique_id()) pair has a
+            // unique Add + Remove pair in the log. For example:
+            // https://github.com/delta-io/delta/blob/master/spark/src/test/resources/delta/table-with-dv-large/_delta_log/00000000000000000001.json
+            if !self.seen.contains(&(add.path.clone(), add.dv_unique_id())) {
+                debug!(
+                    "Including file in scan: ({}, {:?}), is log {is_log_batch}",
+                    add.path,
+                    add.dv_unique_id(),
+                );
+                if is_log_batch {
+                    // Remember file actions from this batch so we can ignore duplicates
+                    // as we process batches from older commit and/or checkpoint files. We
+                    // don't need to track checkpoint batches because they are already the
+                    // oldest actions and can never replace anything.
+                    self.seen.insert((add.path.clone(), add.dv_unique_id()));
+                }
+                selection_vector[index] = true;
+            } else {
+                debug!(
+                    "Filtering out Add due to it being removed {}, is log {is_log_batch}",
+                    add.path
+                );
+            }
+        }
+
+        let result = expression_handler
+            .get_evaluator(
+                get_log_schema().project(&[ADD_NAME])?,
+                self.get_add_transform_expr(),
+                SCAN_ROW_DATATYPE.clone(),
+            )
+            .evaluate(actions)?;
+        Ok((result, selection_vector))
+    }
+
+    // work shared between process_batch and process_scan_batch
+    fn setup_batch_process(
+        &mut self,
+        selection_vector: Option<Vec<bool>>,
+        actions: &dyn EngineData,
+        is_log_batch: bool,
+    ) -> DeltaResult<AddRemoveVisitor> {
+        let schema_to_use = if is_log_batch {
+            // NB: We _must_ pass these in the order `ADD_NAME, REMOVE_NAME` as the visitor assumes
+            // the Add action comes first. The [`project`] method honors this order, so this works
+            // as long as we keep this order here.
+            get_log_schema().project(&[ADD_NAME, REMOVE_NAME])?
+        } else {
+            // All checkpoint actions are already reconciled and Remove actions in checkpoint files
+            // only serve as tombstones for vacuum jobs. So no need to load them here.
+            get_log_schema().project(&[ADD_NAME])?
+        };
+        let mut visitor = AddRemoveVisitor::new(selection_vector, is_log_batch);
+        actions.extract(schema_to_use, &mut visitor)?;
+
+        for remove in visitor.removes.drain(..) {
```

Comment:
Update: It seems like we actually don't reuse the visitor, but we return it so the `removes` vec needs to remain valid. Shouldn't we just return `visitor.adds` instead, since that's all the caller cares about?

Reply:
yeah, that makes more sense :)